### PR TITLE
Multisig wrapper command

### PR DESCRIPTION
**Pull request by RodrigoAD**

### CODE REVIEW

File: packages-ts/gauntlet-terra-cw20-multisig/package.json [hunk 1]
```diff
@@ -0,0 +1,32 @@
+{
+  "name": "@chainlink/gauntlet-terra-cw20-multisig",
```
Comment (krebernisak) [inline packages-ts/gauntlet-terra-cw20-multisig/package.json:2]: Let' s rename this package and have it support the cw-plus contracts we need:

- cw3-flex-multisig
- cw4-group
- cw20-base

```suggestion
  "name": "@chainlink/gauntlet-terra-cw-plus",
```

File: packages-ts/gauntlet-terra/src/commands/internal/terra.ts [hunk 2]
```diff
@@ -63,6 +64,25 @@ export default abstract class TerraCommand extends WriteCommand<TransactionRespo
...
+        }),
+      })
+
+      const res = await this.provider.tx.broadcast(tx)
+
+      logger.debug(res)
+      return this.wrapResponse(res)
+    } catch (e) {
+      const details = e.response.data
+      throw new Error(details.message)
```
Comment (krebernisak) [inline packages-ts/gauntlet-terra/src/commands/internal/terra.ts:82]: What's the purpose of this catch-throw?

Also if there is a reason to leave it in, we should use safe navigation:
```suggestion
      throw new Error(e.response.?data.?message)
```

File: packages-ts/gauntlet-terra/src/commands/internal/terra.ts [hunk 3]
```diff
@@ -63,6 +64,25 @@ export default abstract class TerraCommand extends WriteCommand<TransactionRespo
...
+      const tx = await this.wallet.createAndSignTx({
+        msgs: messages,
+        ...(this.wallet.key instanceof LedgerKey && {
+          signMode: SignMode.SIGN_MODE_LEGACY_AMINO_JSON,
+        }),
+      })
+
+      const res = await this.provider.tx.broadcast(tx)
+
+      logger.debug(res)
```
Comment (krebernisak) [inline packages-ts/gauntlet-terra/src/commands/internal/terra.ts:78]: This debug log looks out of place, do we want it? Will this only appear with `--debug` flag?

File: packages-ts/gauntlet-terra-contracts/src/commands/abstract/inspectionWrapper.ts [hunk 4]
```diff
@@ -47,6 +47,10 @@ export const instructionToInspectCommand = <CommandInput, Expected>(
       super(flags, args)
     }
 
+    makeRawTransaction = () => {
+      throw new Error('Inspection command does not involve any transaction')
```
Comment (krebernisak) [inline packages-ts/gauntlet-terra-contracts/src/commands/abstract/inspectionWrapper.ts:51]: Yeah, read functions should be of a different type.

File: packages-ts/gauntlet-terra-cw20-multisig/src/commands/multisig.ts [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+
+    constructor(flags, args) {
+      super(flags, args)
+
+      this.command = new command(flags, args)
+
+      if (!AccAddress.validate(process.env.MULTISIG_ADDRESS)) throw new Error(`Invalid Multisig wallet address`)
+      if (!AccAddress.validate(process.env.MULTISIG_GROUP)) throw new Error(`Invalid Multisig group address`)
+      this.multisig = process.env.MULTISIG_ADDRESS as AccAddress
+      this.multisigGroup = process.env.MULTISIG_GROUP as AccAddress
```
Comment (krebernisak) [inline packages-ts/gauntlet-terra-cw20-multisig/src/commands/multisig.ts:66]: This should really be read from the input, and not directly from env vars here. Env vars should be completely abstracted from the command. The command just gets some kind of input (flags) that can be loaded from anywhere (CLI flags/args, env, file ...)

File: packages-ts/gauntlet-terra-cw20-multisig/src/commands/multisig.ts [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+      const proposeInput = {
+        propose: {
+          description: command.id,
+          msgs: [this.toWasmMsg(message)],
+          title: command.id,
+          // TODO: Set expiration time
+          // latest: { never: {} }
+        },
+      }
+      return new MsgExecuteContract(signer, this.multisig, proposeInput)
```
Comment (krebernisak) [inline packages-ts/gauntlet-terra-cw20-multisig/src/commands/multisig.ts:118]: Where do we define the wasm function to call? e.g. `begin_proposal`?

File: packages-ts/gauntlet-terra-cw20-multisig/src/commands/multisig.ts [hunk 7]
```diff
@@ -0,0 +1,250 @@
...
+      const approvalInput = {
+        vote: {
+          vote: Vote.YES,
+          proposal_id: proposalId,
+        },
+      }
+      return new MsgExecuteContract(signer, this.multisig, approvalInput)
+    }
+
+    executeExecution: ProposalAction = async (signer, proposalId) => {
```
Comment (krebernisak) [inline packages-ts/gauntlet-terra-cw20-multisig/src/commands/multisig.ts:132]: Execute execution? :D

Comment (krebernisak) [reply packages-ts/gauntlet-terra-cw20-multisig/src/commands/multisig.ts:132]: None of these are "executing" anything, but just building payloads. Consider renaming.

File: packages-ts/gauntlet-terra/src/commands/internal/terra.ts (hunk 2 above)
Comment (RodrigoAD) [reply packages-ts/gauntlet-terra/src/commands/internal/terra.ts:84]: The default `e.message` gives almost no information. Adding the safe navigation

File: packages-ts/gauntlet-terra/src/commands/internal/terra.ts (hunk 3 above)
Comment (RodrigoAD) [reply packages-ts/gauntlet-terra/src/commands/internal/terra.ts:80]: Yeah we don't really need it, removing it

File: packages-ts/gauntlet-terra-cw20-multisig/src/commands/multisig.ts [hunk 8]
```diff
@@ -0,0 +1,275 @@
...
+      const approvalsLeft = state.threshold - state.approvers.length
+      const messages = {
+        [Action.APPROVE]: `The proposal needs ${approvalsLeft} more approvals. Run the same command with the flag --proposal=${proposalId}`,
+        [Action.EXECUTE]: `The proposal reached the threshold and can be executed. Run the same command with the flag --proposal=${proposalId}`,
+        [Action.NONE]: `The proposal has been executed. No more actions needed`,
+      }
+      logger.line()
+      logger.info(`${messages[state.nextAction]}`)
+      logger.line()
+    }
```
Comment (krebernisak) [inline packages-ts/gauntlet-terra-cw20-multisig/src/commands/multisig.ts:198]: This is something we used to call an inspection command.

Here it's just a function, on a different command class. 
The Multisig command, that groups all Multisig abstract commands?

How do I call this inspection command? How do I compose it in a workflow?
Is the Multisig command a workflow or is it something else? What exactly?

I have many questions :)

...

The first priority is to have it function in any way possible, and while this looks ok IMO we need to go back to the Gauntlet v4 drawing board and start typing and composing these commands better.

File: packages-ts/gauntlet-terra-cw20-multisig/src/commands/multisig.ts [hunk 9]
```diff
@@ -0,0 +1,275 @@
...
+      logger.info(`Proposal State:
+        - Total Owners: ${state.owners.length}
+        - Owners List: ${state.owners}
+
+        - Threshold: ${state.threshold}
+        - Total Approvers: ${state.approvers.length}
+        - Approvers List: ${state.approvers}
+
+        - Next Action: ${state.nextAction.toUpperCase()}
+      `)
```
Comment (krebernisak) [inline packages-ts/gauntlet-terra-cw20-multisig/src/commands/multisig.ts:219]: This is another inspection command baked in ad-hoc.

File: packages-ts/gauntlet-terra-cw20-multisig/src/commands/multisig.ts (hunk 7 above)
Comment (RodrigoAD) [reply packages-ts/gauntlet-terra-cw20-multisig/src/commands/multisig.ts:132]: Hehe yes, I'll rename it

File: packages-ts/gauntlet-terra-cw20-multisig/src/commands/multisig.ts (hunk 5 above)
Comment (yosriady) [reply packages-ts/gauntlet-terra-cw20-multisig/src/commands/multisig.ts:66]: Regarding loading environment variables, how about something like this:

```
const environment: Environment = load(.env filepath)

...

executeCLI(environment, myCommands)
```

Then, in each command:

```
this.multisig = this.env.MULTISIG_ADDRESS
```

where `this.env` contains the environment variables, available in each `Command` instance.

File: packages-ts/gauntlet-terra-cw20-multisig/src/commands/multisig.ts (hunk 6 above)
Comment (RodrigoAD) [reply packages-ts/gauntlet-terra-cw20-multisig/src/commands/multisig.ts:118]: Not sure is this is what you mean, but here we decide which operation to call: https://github.com/smartcontractkit/chainlink-terra/pull/111/files#diff-c7c2eafb6b93bd0e456c2fd133626df8051e82da80db2ad3695f45fe6cbd72bbR88
That's based on `state.nextAction`

File: packages-ts/gauntlet-terra-cw20-multisig/src/commands/multisig.ts (hunk 8 above)
Comment (RodrigoAD) [reply packages-ts/gauntlet-terra-cw20-multisig/src/commands/multisig.ts:198]: I see your point. This goes a bit beyond from this PR. Opening an issue for it: #133

File: packages-ts/gauntlet-terra-cw20-multisig/src/commands/multisig.ts (hunk 9 above)
Comment (RodrigoAD) [reply packages-ts/gauntlet-terra-cw20-multisig/src/commands/multisig.ts:219]: #133

File: packages-ts/gauntlet-terra-cw20-multisig/src/commands/multisig.ts (hunk 5 above)
Comment (RodrigoAD) [reply packages-ts/gauntlet-terra-cw20-multisig/src/commands/multisig.ts]: Opening an issue here, thought could be a simpler fix, but Gauntlet is not really prepared to inject env vars anywhere.
https://github.com/smartcontractkit/chainlink-terra/issues/134

**Pull request merged**